packages/@lwc/engine-dom/src/apis/build-custom-element-constructor.ts

```diff
@@ -59,26 +55,19 @@ export function buildCustomElementConstructor(Ctor: ComponentConstructor): HTMLE
     return class extends HTMLElement {
         constructor() {
             super();
-
-            if (this.isConnected) {
-                // this if block is hit when there's already an un-upgraded element in the DOM with the same tag name.
-                hydrateComponent(this, Ctor, {});
-                hydratedCustomElements.add(this);
-            } else {
-                createVM(this, Ctor, renderer, {
-                    mode: 'open',
-                    owner: null,
-                    tagName: this.tagName,
-                });
-            }
+            assert.isTrue(
+                isNull(this.shadowRoot),
+                `Found an existing shadow root for the custom element ${Ctor.name}. Call \`hydrateComponent\` instead.`
+            );
+            assert.isTrue(this.childNodes.length == 0, `Top level elements cannot have children.`);
```

Comment:
Does it make sense to error here, or should we just log a warning and move on?

I'm wondering about cases where someone does something like this:

```html
<x-foo>
</x-foo>
```

This appears empty, but because it contains whitespace, it technically has child nodes (text nodes). Same with:

```html
<x-foo><!-- lalala --></x-foo>
```

Also, I'm just worried about making too drastic of a breaking change.

Reply:
That's a fair point. For the existing shadow root, I think we should throw an error. But if there are children, perhaps we just remove them instead of throwing. I don't really prefer one behavior over another - I think they're both fine - but warning & removing children is less likely to be a disruptive breaking change.